src/components/tl/mlx5/mcast/tl_mlx5_mcast.h

```diff
@@ -95,24 +150,145 @@ typedef struct ucc_tl_mlx5_mcast_context {
+    ucc_list_link_t                super;
+    int                            pkt_id;
+    ucc_tl_mlx5_mcast_coll_comm_t *comm;
+} ucc_tl_mlx5_mcast_nack_req_t;
+
+#define PSN_IS_IN_RANGE(_psn, _call, _comm)                                         \
+                             (                                                      \
+                                ((_psn >= _call->start_psn) &&                      \
+                                (_psn < _call->start_psn + _call->num_packets) &&   \
+                                (_psn >= _comm->last_acked) &&                      \
+                                (_psn < _comm->last_acked + _comm->wsize))          \
+                             )
+
+#define PSN_TO_RECV_OFFSET(_psn, _call, _comm)                                      \
+                             (                                                      \
+                                ((ptrdiff_t)((_psn - _call->start_psn)              \
+                                 * (_comm->max_per_packet)))                        \
+                             )
+
+#define PSN_TO_RECV_LEN(_psn, _call, _comm)                                         \
+                             (                                                      \
+                                ((_psn - _call->start_psn + 1) %                    \
+                                 _call->num_packets == 0 ? _call->last_pkt_len :    \
+                                 _comm->max_per_packet)                             \
+                             )
+
+#define PSN_RECEIVED(_psn, _comm)                                                   \
+                             (                                                      \
+                                (_comm->r_window[(_psn) %                           \
+                                 _comm->wsize]->psn == (_psn))                      \
+                             )
+
+typedef struct ucc_tl_mlx5_mcast_coll_req {
+    ucc_tl_mlx5_mcast_coll_comm_t  *comm;
+    size_t                          length; /* bcast buffer size */
+    int                             proto;
+    struct ibv_mr                  *mr;
+    struct ibv_recv_wr             *rwr;
+    struct ibv_sge                 *rsgs;
+    void                           *rreg;
+    char                           *ptr;
+    int                             am_root;
+    ucc_rank_t                      root;
+    void                          **rbufs;
+    int                             first_send_psn;
+    int                             to_send;
+    int                             to_recv;
+    ucc_rank_t                      parent;
+    uint32_t                        start_psn;
+    int                             num_packets;
+    int                             last_pkt_len;
+    int                             offset;
+    ucc_memory_type_t               buf_mem_type;
 } ucc_tl_mlx5_mcast_coll_req_t;
```

Comment:
why do you need to duplication length, buf_mem_type, ptr etc? It's available in coll_task args